File: bench/langs/nrev.py
"""naive reverse -- the same algorithm, cell for cell.

`app` is the quadratic append Prolog's nrev is built from; keeping it
means both sides do the same work rather than the same errand.
"""
import sys

def app(a, b):
    r = list(b)
    for x in reversed(a):
        r.insert(0, x)
    return r

def nrev(l):
    if not l:
        return []
    return app(nrev(l[1:]), [l[0]])

def sum31(l):
    a = 0
    for h in l:
        a = (a * 31 + h) % 1000003
    return a

def main(n, reps):
    l = list(range(1, n + 1))
    c = 0
    for _ in range(reps):
        c = sum31(nrev(l))
    print("answer(%d)" % c)

if __name__ == "__main__":
    sys.setrecursionlimit(100000)
    main(int(sys.argv[1]), int(sys.argv[2]))

File: bench/langs/nrev.pl
%% naive reverse -- the classic Prolog benchmark, and a list-building
%% workload for anything else. `app/3' is quadratic on purpose: this
%% measures the cost of building and matching list cells, which is what a
%% Prolog engine is made of.
%%
%%   main(+N, +Reps)   reverse [1..N] Reps times; answer is a checksum

app([], L, L).
app([H|T], L, [H|R]) :- app(T, L, R).

nrev([], []).
nrev([H|T], R) :- nrev(T, RT), app(RT, [H], R).

sum31([], A, A).
sum31([H|T], A0, A) :- A1 is (A0 * 31 + H) mod 1000003, sum31(T, A1, A).

%% Every rep is the same unit of work and answers the same checksum, so
%% the ANSWER does not depend on how many reps the harness chose -- which
%% is what lets the two languages run different rep counts (each long
%% enough to be a measurement) and still be checked against each other.
rounds(0, _, C, C) :- !.
rounds(K, L, _, C) :-
    nrev(L, R), sum31(R, 0, C0),
    K1 is K - 1,
    rounds(K1, L, C0, C).

main(N, Reps) :-
    numlist(1, N, L),
    rounds(Reps, L, 0, C),
    format("answer(~w)~n", [C]).
